pkg/deploy/strategy/support/lifecycle.go

```diff
@@ -239,100 +239,75 @@ func NewFirstContainerReady(kclient kclient.Interface, timeout time.Duration, in
-	// TODO: should finding multiple be an error?
-	pod := &pods.Items[0]
-	if len(pods.Items) > 1 {
-		glog.Infof("Warning: more than one pod for deployment %s; basing canary check on the first pod '%s'", deployutil.LabelForDeployment(deployment), pod.Name)
-	}
-
+func (c *AcceptNewlyObservedReadyPods) Accept(deployment *kapi.ReplicationController) error {
 	// Make a pod store to poll and ensure it gets cleaned up.
-	podStore, stopStore := c.getPodStore(pod.Namespace, pod.Name)
+	podStore, stopStore := c.getDeploymentPodStore(deployment)
 	defer close(stopStore)
 
-	// Track container readiness based on those defined in the spec.
-	observedContainers := map[string]bool{}
-	for _, container := range pod.Spec.Containers {
-		observedContainers[container.Name] = false
-	}
-
 	// Start checking for pod updates.
-	glog.V(0).Infof("Waiting for pod %s/%s container readiness", pod.Namespace, pod.Name)
-	err = wait.Poll(c.interval, c.timeout, func() (done bool, err error) {
-		// Get the latest state of the pod.
-		obj, exists, err := podStore.Get(pod)
-		// Try again later on error or if the pod isn't available yet.
-		if err != nil {
-			glog.V(0).Infof("Error getting pod %s/%s to inspect container readiness: %v", pod.Namespace, pod.Name, err)
-			return false, nil
-		}
-		if !exists {
-			glog.V(0).Infof("Couldn't find pod %s/%s to inspect container readiness", pod.Namespace, pod.Name)
-			return false, nil
-		}
-		// New pod state is available; update the observed ready status of any
-		// containers.
-		updatedPod := obj.(*kapi.Pod)
-		for _, status := range updatedPod.Status.ContainerStatuses {
-			// Ignore any containers which aren't defined in the deployment spec.
-			if _, known := observedContainers[status.Name]; !known {
-				glog.V(0).Infof("Ignoring readiness of container %s in pod %s/%s because it's not present in the pod spec", status.Name, pod.Namespace, pod.Name)
+	glog.V(0).Infof("Waiting %.f seconds for pods owned by deployment %q to become ready (checking every %.f seconds; %d pods previously accepted)", c.timeout.Seconds(), deployutil.LabelForDeployment(deployment), c.interval.Seconds(), c.acceptedPods.Len())
+	err := wait.Poll(c.interval, c.timeout, func() (done bool, err error) {
+		// Check for pod readiness.
+		unready := kutil.NewStringSet()
+		for _, obj := range podStore.List() {
+			pod := obj.(*kapi.Pod)
+			// Skip previously accepted pods; we only want to verify newly observed
+			// and unaccepted pods.
+			if c.acceptedPods.Has(pod.Name) {
 				continue
 			}
-			// The status of the container could be transient; we only care if it
-			// was ever ready. If it was ready and then became not ready, we
-			// consider it ready.
-			if status.Ready {
-				observedContainers[status.Name] = true
-			}
-		}
-		// Check whether all containers have been observed as ready.
-		allReady := true
-		for _, ready := range observedContainers {
-			if !ready {
-				allReady = false
-				break
+			if kapi.IsPodReady(pod) {
```

Comment:
I think the latter.  But you can't scale down if one of your new ones goes
down without breaching SLA.

On Jul 21, 2015, at 12:09 PM, Michail Kargakis <notifications@github.com>
wrote:

In pkg/deploy/strategy/support/lifecycle.go
<https://github.com/openshift/origin/pull/3701#discussion_r35120393>:

>  				continue
>  			}
> -			// The status of the container could be transient; we only care if it
> -			// was ever ready. If it was ready and then became not ready, we
> -			// consider it ready.
> -			if status.Ready {
> -				observedContainers[status.Name] = true
> -			}
> -		}
> -		// Check whether all containers have been observed as ready.
> -		allReady := true
> -		for _, ready := range observedContainers {
> -			if !ready {
> -				allReady = false
> -				break
> +			if kapi.IsPodReady(pod) {

If the latter, make this

if c.acceptedPods.Has(pod) && kapi.IsPodReady(pod) {

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3701/files#r35120393>.